### ISSUE TITLE
NoMethodError for ObjectSpace::WeakMap#values

### ISSUE THREAD

**Issue by yahonda** [jruby/jruby#6267]
**Environment Information**
  
* JRuby version (`jruby -v`) and command line (flags, JRUBY_OPTS, etc)
```
% jruby -v
jruby 9.2.11.1 (2.5.7) 2020-03-25 b1f55b1a40 OpenJDK 64-Bit Server VM 11.0.2+9 on 11.0.2+9 +jit [darwin-x86_64]
```

* Operating system and platform (e.g. `uname -a`)
```
% uname -a
Darwin mymacbookpro 19.6.0 Darwin Kernel Version 19.6.0: Sun May 17 22:15:23 PDT 2020; root:xnu-6153.140.21~10/RELEASE_X86_64 x86_64
```

**Expected Behavior**

* Steps to reproduce
```ruby
weakmap = ObjectSpace::WeakMap.new
weakmap.values
```

`weakmap.values` should return empty Array as CRuby does.

* Result with CRuby 2.7.1

```
% ruby -v
ruby 2.7.1p83 (2020-03-31 revision a0c7c23c9c) [x86_64-darwin19]
% irb
irb(main):001:0> weakmap = ObjectSpace::WeakMap.new
irb(main):002:0> weakmap.values
=> []
irb(main):003:0>
```

**Actual Behavior**

It raises `NoMethodError`

```
% ruby -v
jruby 9.2.11.1 (2.5.7) 2020-03-25 b1f55b1a40 OpenJDK 64-Bit Server VM 11.0.2+9 on 11.0.2+9 +jit [darwin-x86_64]
% irb
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.headius.backport9.modules.Modules to method sun.nio.ch.NativeThread.signal(long)
WARNING: Please consider reporting this to the maintainers of com.headius.backport9.modules.Modules
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
irb(main):001:0> weakmap = ObjectSpace::WeakMap.new
=> #<ObjectSpace::WeakMap:0x189b5fb1>
irb(main):002:0> weakmap.values
Traceback (most recent call last):
        6: from /Users/yahonda/.rbenv/versions/jruby-9.2.11.1/bin/irb:13:in `<main>'
        5: from org/jruby/RubyKernel.java:1189:in `catch'
        4: from org/jruby/RubyKernel.java:1189:in `catch'
        3: from org/jruby/RubyKernel.java:1442:in `loop'
        2: from org/jruby/RubyKernel.java:1048:in `eval'
        1: from (irb):2:in `evaluate'
NoMethodError (undefined method `values' for #<ObjectSpace::WeakMap:0x189b5fb1>)
irb(main):003:0>
```

**Comment by yahonda** [comment]
This is required to support https://github.com/rsim/oracle-enhanced/issues/2027

**Comment by yahonda** [comment]
Let me update that his issue needs to be addressed to support https://github.com/rails/rails/pull/39121 .

**Comment by headius** [comment]
Should be easy enough to support.

**Comment by headius** [comment]
Ok, so I don't have a problem with adding the missing features from Weakmap, but officially *it is not a public API*. It has never been discussed (naming, behavior, etc) and I believe that needs to happen before it sees usage in third-party libraries. Unfortunately that would mean this should not be used until after Ruby 2.8/3.0, though JRuby would probably implement everything that CRuby provides in our 9.3 release.

https://bugs.ruby-lang.org/issues/16959

I'd like to get some feedback on this before we proceed to implement the rest of Weakmap just because CRuby does it that way.

**Comment by headius** [comment]
FWIW I don't fault you for using the feature... it's sorely needed in Ruby, as I've pointed out in my ReferenceQueue issue (linked from the Ruby issue above), and now that there's specs it's hard to tell that it's not a public API.

**Comment by yahonda** [comment]
Thanks for the update. I with `ObjectSpace::WeakMap#values` will be implemented for JRuby.

**Comment by headius** [comment]
FWIW this can be implemented in Ruby using some JRuby internal APIs (non-public, but useful for cases like this):

```ruby
require 'jruby'

class org::jruby::RubyObjectSpace::WeakMap
  field_reader :map
end

class ObjectSpace::WeakMap
  def values
    JRuby.ref(self).map.values.reject(&:nil?)
  end
end

# Example usage follows...

wm = ObjectSpace::WeakMap.new

x = nil

# This is inside a block to avoid references leaking to the main scope
tap {
  wm["foo"] = Object.new
  wm["bar"] = x = Object.new

  p wm.values
}

# $ jruby weakmap_values.rb
# [#<Object:0x62e136d3>, #<Object:0x51b7e5df>]

5.times { JRuby.gc }

p wm.values

# $ jruby weakmap_values.rb
# [#<Object:0x62e136d3>]
```

The `reject` is necessary because the data structure in `map` does not omit vacated references (I assume the CRuby implementation does this filtering interally too).

**Comment by ryudoawaru** [comment]
Hi @headius 

I am trying using newest [oracle-enhanced](https://github.com/rsim/oracle-enhanced) adapter for rails.

I found it seems that the patch mentioned in this issue are not necessary anymore right?

Since current main branch of oracle-enhanced are still with these patches, I found the [patch](https://github.com/rsim/oracle-enhanced/blob/master/lib/active_record/connection_adapters/oracle_enhanced_adapter.rb#L850) of adding field_reader to class `org.jruby::RubyObjectSpace::WeakMap`  would raise error when entering rails console.

```
NameError: undefined field 'map' for class 'Java::OrgJruby::RubyObjectSpace::WeakMap'
            field_reader at org/jruby/java/proxies/JavaProxy.java:352
```

After I remove the patch, it turns to all right now.

So I want to send a PR to the `oracle-enhanced` and just want to ask if  it is still necessary to add `#values` method [patch](https://github.com/rsim/oracle-enhanced/blob/master/lib/active_record/connection_adapters/oracle_enhanced_adapter.rb#L854) now?

**Comment by headius** [comment]
@ryudoawaru Good discovery! Yes I think you should file a PR to oracle-enhanced and let them know this is no longer an issue. This was fixed when I added all remaining missing WeakMap methods in https://github.com/jruby/jruby/pull/6683.

**Comment by redconfetti** [comment]
I had to remove this workaround from our fork of rsim/oracle-enhanced. I'm using the [release80 branch](https://github.com/rsim/oracle-enhanced/tree/release80), with fix from rsim/oracle-enhanced#2473 applied, and this workaround removed, to get our app to work with Rails v8.0.4.

I've got this our [fork branch](https://github.com/sis-berkeley-edu/oracle-enhanced/tree/release80_plus_2473).
